tonic/src/transport/server/incoming.rs

```diff
@@ -164,33 +164,66 @@ impl TcpIncoming {
     /// # }
     pub fn new(
         addr: SocketAddr,
-        nodelay: bool,
-        keepalive: Option<Duration>,
+        tcp_nodelay: bool,
+        tcp_keepalive_timeout: Option<Duration>,
     ) -> Result<Self, crate::Error> {
-        let mut inner = AddrIncoming::bind(&addr)?;
-        inner.set_nodelay(nodelay);
-        inner.set_keepalive(keepalive);
-        Ok(TcpIncoming { inner })
+        let std_listener = StdTcpListener::bind(addr)?;
+        std_listener.set_nonblocking(true)?;
+        let inner = TcpListenerStream::new(TcpListener::from_std(std_listener)?);
+        Ok(Self {
+            inner,
+            tcp_nodelay,
+            tcp_keepalive_timeout,
+        })
     }
 
     /// Creates a new `TcpIncoming` from an existing `tokio::net::TcpListener`.
     pub fn from_listener(
         listener: TcpListener,
-        nodelay: bool,
-        keepalive: Option<Duration>,
+        tcp_nodelay: bool,
+        tcp_keepalive_timeout: Option<Duration>,
     ) -> Result<Self, crate::Error> {
-        let mut inner = AddrIncoming::from_listener(listener)?;
-        inner.set_nodelay(nodelay);
-        inner.set_keepalive(keepalive);
-        Ok(TcpIncoming { inner })
+        Ok(Self {
+            inner: TcpListenerStream::new(listener),
+            tcp_nodelay,
+            tcp_keepalive_timeout,
+        })
     }
 }
 
 impl Stream for TcpIncoming {
-    type Item = Result<AddrStream, std::io::Error>;
+    type Item = Result<TcpStream, std::io::Error>;
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        Pin::new(&mut self.inner).poll_accept(cx)
+        match ready!(Pin::new(&mut self.inner).poll_next(cx)) {
+            Some(Ok(stream)) => {
+                set_accept_socketoptions(&stream, self.tcp_nodelay, self.tcp_keepalive_timeout);
```

Comment:
Actually, internally, hyper used to set these options on the TcpStream [after accept via AddrIncoming](https://github.com/hyperium/hyper/blob/dedcb674f35eaec765a42b550caabe6f694d86d1/src/server/tcp.rs#L220). I think that isn't exposed because you can't set `nodelay` or `keepalive` on a listening socket because it can't keep alive or delay sending packets to a peer it doesn't yet know about.